lib/Platform/Intl/PlatformIntlICU.cpp

```diff
@@ -120,40 +419,782 @@ double Collator::compare(
+
+  // https://402.ecma-international.org/8.0/#sec-properties-of-intl-datetimeformat-instances
+  // Intl.DateTimeFormat instances have an [[InitializedDateTimeFormat]]
+  // internal slot.
+  // NOTE: InitializedDateTimeFormat is not implemented.
+  // Intl.DateTimeFormat instances also have several internal
+  // slots that are computed by the constructor:
+  // [[Locale]] is a String value with the language tag of the locale whose
+  // localization is used for formatting.
+  std::u16string locale_;
+  // [[Calendar]] is a String value with the "type" given in Unicode Technical
+  // Standard 35 for the calendar used for formatting.
+  std::optional<std::u16string> calendar_;
+  // [[NumberingSystem]] is a String value with the "type" given in Unicode
+  // Technical Standard 35 for the numbering system used for formatting.
+  // NOTE: Even though NSDateFormatter formats date and time using different
+  // numbering systems based on its "locale" value, it does not allow to set/get
+  // the numbering system value explicitly. So we consider this feature
+  // unsupported.
+  // [[TimeZone]] is a String value with the IANA time zone name of the time
+  // zone used for formatting.
+  std::u16string timeZone_;
+  // [[Weekday]], [[Era]], [[Year]], [[Month]], [[Day]], [[DayPeriod]],
+  // [[Hour]], [[Minute]], [[Second]], [[TimeZoneName]] are each either
+  // undefined, indicating that the component is not used for formatting, or one
+  // of the String values given in Table 4, indicating how the component should
+  // be presented in the formatted output.
+  std::optional<std::u16string> weekday_;
+  std::optional<std::u16string> era_;
+  std::optional<std::u16string> year_;
+  std::optional<std::u16string> month_;
+  std::optional<std::u16string> day_;
+  std::optional<std::u16string> dayPeriod_;
+  std::optional<std::u16string> hour_;
+  std::optional<std::u16string> minute_;
+  std::optional<std::u16string> second_;
+  std::optional<std::u16string> timeZoneName_;
+  // [[FractionalSecondDigits]] is either undefined or a positive, non-zero
+  // integer Number value indicating the fraction digits to be used for
+  // fractional seconds. Numbers will be rounded or padded with trailing zeroes
+  // if necessary.
+  std::optional<uint8_t> fractionalSecondDigits_;
+  // [[HourCycle]] is a String value indicating whether the 12-hour format
+  // ("h11", "h12") or the 24-hour format ("h23", "h24") should be used. "h11"
+  // and "h23" start with hour 0 and go up to 11 and 23 respectively. "h12" and
+  // "h24" start with hour 1 and go up to 12 and 24. [[HourCycle]] is only used
+  // when [[Hour]] is not undefined.
+  std::optional<std::u16string> hourCycle_;
+  // [[DateStyle]], [[TimeStyle]] are each either undefined, or a String value
+  // with values "full", "long", "medium", or "short".
+  std::optional<std::u16string> dateStyle_;
+  std::optional<std::u16string> timeStyle_;
+  std::string locale8_;
+  UDateFormat *dateTimeFormatter_;
```

Comment:
Please add a doc-comment for these last two fields.

It also seems like `dateTimeFormatter_` here is being leaked? We should have a destructor for `DateTimeFormatICU` that does the necessary cleanup.